pkg/sources/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,270 @@
+	jobId          sources.JobID
+	concurrency    int
+	verify         bool
+	esConfig       es.Config
+	filterParams   FilterParams
+	bestEffortScan bool
+	ctx            context.Context
+	client         *es.TypedClient
+	log            logr.Logger
+	sources.Progress
+}
+
+// Init returns an initialized Elasticsearch source
+func (s *Source) Init(
+	aCtx context.Context,
+	name string,
+	jobId sources.JobID,
+	sourceId sources.SourceID,
+	verify bool,
+	connection *anypb.Any,
+	concurrency int,
+) error {
+	var conn sourcespb.Elasticsearch
+	if err := anypb.UnmarshalTo(connection, &conn, proto.UnmarshalOptions{}); err != nil {
+		return errors.WrapPrefix(err, "error unmarshalling connection", 0)
+	}
+
+	s.name = name
+	s.sourceId = sourceId
+	s.jobId = jobId
+	s.concurrency = concurrency
+	s.verify = verify
+
+	s.ctx = aCtx
+	s.log = aCtx.Logger()
+
+	esConfig := es.Config{}
+
+	if len(conn.Nodes) > 0 {
+		esConfig.Addresses = conn.Nodes
+	}
+
+	if conn.Username != "" {
+		esConfig.Username = conn.Username
+	}
+
+	if conn.Password != "" {
+		esConfig.Password = conn.Password
+	}
+
+	if conn.CloudId != "" {
+		esConfig.CloudID = conn.CloudId
+	}
+
+	if conn.ApiKey != "" {
+		esConfig.APIKey = conn.ApiKey
+	}
+
+	if conn.ServiceToken != "" {
+		esConfig.ServiceToken = conn.ServiceToken
+	}
+
+	s.esConfig = esConfig
+
+	if conn.IndexPattern == "" {
+		s.filterParams.indexPattern = "*"
+	} else {
+		s.filterParams.indexPattern = conn.IndexPattern
+	}
+
+	s.filterParams.queryJSON = conn.QueryJson
+	s.filterParams.sinceTimestamp = conn.SinceTimestamp
+
+	s.bestEffortScan = conn.BestEffortScan
+
+	client, err := s.buildElasticClient()
+	if err != nil {
+		return err
+	}
+
+	s.client = client
+
+	return nil
+}
+
+func (s *Source) buildElasticClient() (*es.TypedClient, error) {
+	return es.NewTypedClient(s.esConfig)
+}
+
+func (s *Source) Type() sourcespb.SourceType {
+	return SourceType
+}
+
+func (s *Source) SourceID() sources.SourceID {
+	return s.sourceId
+}
+
+func (s *Source) JobID() sources.JobID {
+	return s.jobId
+}
+
+// Chunks emits chunks of bytes over a channel.
+func (s *Source) Chunks(
+	ctx context.Context,
+	chunksChan chan *sources.Chunk,
+	targets ...sources.ChunkingTarget,
+) error {
+	indices := Indices{filterParams: &s.filterParams}
+
+	for {
+		workerPool := new(errgroup.Group)
+		workerPool.SetLimit(s.concurrency)
+
+		previousDocumentCount := indices.documentCount
+		err := indices.Update(s.ctx, s.client)
+		if err != nil {
+			return err
+		}
+
+		// The scanCoverageRate is documentsScanned / documentsAdded. If it's not 1,
+		// we need each DocumentSearch to skip some records.
+		scanCoverageRate := 1.0
+		if previousDocumentCount > 0 && indices.documentCount > 0 {
+			scanCoverageRate =
+				float64(previousDocumentCount) / float64(indices.documentCount)
+		}
+
+		unitsOfWork := distributeDocumentScans(&indices, s.concurrency, scanCoverageRate)
+
+		for uowIndex, outerUOW := range unitsOfWork {
+			uow := outerUOW
```

Comment:
@camgunz 
**Suggestion**: Even though `uowIndex` is only used in logs, it still needs to have a copy similar to `uow` inside the loop otherwise we have the potential for incorrect log output

Reply:
Oof, nice catch; I think this started life as `_` but then I used it for logging and missed pulling it in 👍